File: Solution/Programmers/합승 택시 요금.py
import heapq
INF = int(1e9)

def dijkstra(distance, start, graph):
    q = []
    heapq.heappush(q, (0, start))
    distance[start] = 0
    while q:
        dist, curr = heapq.heappop(q)
        if distance[curr] < dist:
            continue
        for i in graph[curr]:
            cost = dist + i[1]
            if cost < distance[i[0]]:
                distance[i[0]] = cost
                heapq.heappush(q, (cost, i[0]))

def solution(n, s, a, b, fares):
    graph = [[] for _ in range(n+1)]
    for fare in fares:
        x, y, z = fare
        graph[x].append((y, z))
        graph[y].append((x, z))
    
    distance = [INF] * (n+1)
    dijkstra(distance, s, graph)
    ans = INF
    for i in range(1, n+1):
        result = 0
        distance_tmp = [INF] * (n+1)
        dijkstra(distance_tmp, i, graph)
        result = distance[i] + distance_tmp[a] + distance_tmp[b]
        ans = min(ans, result)
    return ans

File: Solution/Programmers/합승 택시 요금.js
function solution(n, s, a, b, fares) {
  const graph = Array.from(Array(n + 1), () => Array(n + 1).fill(Infinity));

  fares.forEach(fare => {
    const [c, d, f] = fare;
    graph[c][d] = f;
    graph[d][c] = f;
  });

  for (let k = 1; k < n + 1; k++) {
    for (let i = 1; i < n + 1; i++) {
      for (let j = 1; j < n + 1; j++) {
        if (i === j) graph[i][j] = 0;
        if (graph[i][k] + graph[k][j] < graph[i][j]) {
          graph[i][j] = graph[i][k] + graph[k][j];
        }
      }
    }
  }

  let ans = graph[s][a] + graph[s][b];
  for (let i = 1; i < n + 1; i++) {
    ans = Math.min(ans, graph[s][i] + graph[i][a] + graph[i][b]);
  }

  return ans;
}
